# Regression Inference

In [1]:
%load_ext pretty_jupyter 

## Standard Inference

On this page, you will learn how to implement regression inference procedures using Python. Before we begin, we need to import the necessary Python libraries. Unlike R, Python requires explicit package installation for data handling and statistical modelling. For this workthrough, we'll use `Pandas` for data manipulation, `Statsmodels` for statistical modelling, and `lets-plot` for visualization.

In [3]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
from lets_plot import *

### Loading and Preparing the Data
We'll use the Women’s Wages dataset ([mroz.csv](https://github.com/datasquad/ECLR/blob/gh-pages/data/Mroz.csv())). First, ensure that your dataset is in the correct directory or update the path accordingly. Now, make sure that the file location and working directory are synchronised. The data file we practice with here is a Comma-Separated Values (CSV) file. To load such a file in Python, we use the`read_csv` function from the `pandas` library, assuming that the data file is in the working directory:

In [5]:
import os
print("Current Working Directory:", os.getcwd())

Current Working Directory: C:\Users\Admin


In [7]:
os.chdir("C:/Users/Admin/Desktop/Pycode/ECLR/data")  # Replace with your desired path
print("New Working Directory:", os.getcwd())

New Working Directory: C:\Users\Admin\Desktop\Pycode\ECLR\data


Now, we are ready to import the data.

In [9]:
# Load the CSV file
mydata = pd.read_csv("Mroz.csv")

Ensure all variables used in the regression are numeric:

In [11]:
print(mydata.dtypes)

inlf          int64
hours         int64
kidslt6       int64
kidsge6       int64
age           int64
educ          int64
wage         object
repwage     float64
hushrs        int64
husage        int64
huseduc       int64
huswage     float64
faminc        int64
mtr         float64
motheduc      int64
fatheduc      int64
unem        float64
city          int64
exper         int64
nwifeinc    float64
lwage        object
expersq       int64
dtype: object


If any variables are not numeric (e.g., boolean, string), convert them to numeric types if appropriate. For instance, we can see that `wage` and `wage` are **objects**.  This suggests these columns might contain non-numeric data, possibly strings or mixed types (like numbers and strings). This can happen if the data contains symbols or letters or if there are issues during the data import process ( for more information refer back to the [Loading Data in Python](https://datasquad.github.io/ECLR/Py_coding/Py_LoadingData.html) section in the ECLR. To  handle and convert these columns to a numeric format for analysis, you will need to follow specific steps:

[//]: # (-.- .tabset)

#### Step 1: Check the Unique Values

First, let's see what kind of data these columns contain. This will help us understand why they're recognized as `object` types.

In [13]:
print(mydata['wage'].unique())
print(mydata['lwage'].unique())

['3.354' '1.3889' '4.5455' '1.0965' '4.5918' '4.7421' '8.3333' '7.8431'
 '2.1262' '4.6875' '4.063' '2.0833' '2.2668' '3.6797' '1.3472' '3.2143'
 '5.175' '2' '7.5529' '3.5052' '3.5714' '3.25' '2.1545' '3.7879' '4'
 '4.7269' '7.2559' '5.8671' '1.5385' '2.459' '5.8511' '3.8068' '2.4638'
 '2.3753' '4.5351' '5.6183' '14.631' '2.6786' '3.9194' '2.5729' '4.5375'
 '3.4722' '2.0161' '4.5716' '2.2727' '2.6375' '2.2899' '1.0989' '1.1765'
 '1.6' '1.8762' '4.0437' '9.6354' '8.0409' '4.599' '2.1429' '4.4' '3.5354'
 '2.7174' '6.25' '11.933' '3.5931' '6.9444' '2.9167' '3.0769' '3.75'
 '5.7259' '3.6757' '5.1648' '8.2237' '4.3365' '4.9819' '0.3571' '2.9674'
 '1' '2.5554' '0.8602' '2.9261' '3.5461' '1.6264' '3.0952' '2.7' '5.2521'
 '1.4154' '4.7986' '1.6667' '1.1217' '0.5' '0.7143' '2.7961' '4.8583'
 '1.7435' '2.4631' '2.4213' '1.5345' '2.8818' '2.4069' '5.2326' '3.2313'
 '3.4014' '1.3333' '9.3023' '4.5' '4.6235' '3.9557' '5.814' '4.0816' '6'
 '3.6667' '3.8613' '2.7629' '2.931' '4.3884' '5.4167' '9.8611'

#### Step 2: Convert to Numeric
Once you've identified and possibly cleaned any non-numeric entries, you can convert these columns to numeric. Pandas' `to_numeric` function is useful here, particularly because it can handle errors and convert problematic entries to `NaN` (which can then be addressed separately).

In [15]:
# Convert 'wage' and 'lwage' to numeric, setting errors='coerce' will convert non-convertible terms to NaN
mydata['wage'] = pd.to_numeric(mydata['wage'], errors='coerce')
mydata['lwage'] = pd.to_numeric(mydata['lwage'], errors='coerce')

#### Step 3: Handling Missing Values
After conversion, check for and decide how to handle any `NaN` values that may have been introduced during this process.

In [17]:
# Check how many NaN values were introduced
print(mydata['wage'].isna().sum())
print(mydata['lwage'].isna().sum())

325
325


Let's drops these rows.

In [19]:
# Drop these rows
mydata.dropna(subset=['wage', 'lwage'], inplace=True)  # Dropping rows with NaN values

#### Step 4: Verify the Conversion
Ensure that the conversion has been successful and that the data types are now numeric.

In [21]:
print(mydata['wage'].dtype)
print(mydata['lwage'].dtype)

float64
float64


Now we can indicate the the data type for each of the `wage` and `lwage` columns are now a float!

In [23]:
print(mydata.dtypes)

inlf          int64
hours         int64
kidslt6       int64
kidsge6       int64
age           int64
educ          int64
wage        float64
repwage     float64
hushrs        int64
husage        int64
huseduc       int64
huswage     float64
faminc        int64
mtr         float64
motheduc      int64
fatheduc      int64
unem        float64
city          int64
exper         int64
nwifeinc    float64
lwage       float64
expersq       int64
dtype: object


[//]: # (-.- .unlisted .unnumbered)

### Simple OLS Regression Model

To construct a foundational model, we perform an Ordinary Least Squares (OLS) regression to predict `lwage` using the predictors: a constant term, experience, and the logarithm of the husband's wage (`log(huswage)`). The model's equation is:

$$lwage_i = β_0 + β_1 exper_i + β_2 log(huswage_i) + u_i$$

Let's first ensure that we have the log-transformed `huswage`. This step involves ensuring that all values in `huswage` are positive. Logarithms are not defined for zero or negative values. If your dataset might contain zero or negative wages, you need to handle these cases—either by filtering them out or adjusting them accordingly!

[//]: # (-.- .tabset)

#### Step 1: Ensure Positive Values

In [25]:
# Filter out rows where huswage is less than or equal to zero
mydata = mydata[mydata['huswage'] > 0]

You can then compute the logarithm of the `huswage` column using `numpy's` log function. This will apply the natural logarithm (base e) to each value in the column:


#### Step 3: Compute the Logarithm

In [27]:
mydata['log_huswage'] = np.log(mydata['huswage'])

If you expect to have zero or negative values and cannot filter them out, you might need to adjust them before applying the logarithm. One common approach is to add a small constant to all values to ensure they are positive:

#### Step 4: Handling Missing or Zero Values

In [29]:
# Adding a small constant, for example, 1, to avoid log(0) which is undefined
mydata['log_huswage'] = np.log(mydata['huswage'] + 1)

This approach is often used in situations where zero represents a meaningful value (like absence of income) that you want to retain in your analysis. Adjusting by a small constant like 1 makes the logarithmic transformation feasible while minimizing the distortion to the data.

#### Step 5: Verify the Transformation

After applying the transformation, it's a good practice to check the new column to ensure the transformation was applied correctly:

In [31]:
# Check the first few entries of the new log-transformed column
print(mydata[['huswage', 'log_huswage']].head())

   huswage  log_huswage
0   4.0288     1.615181
1   8.4416     2.245125
2   3.5807     1.521852
3   3.5417     1.513301
4  10.0000     2.397895


[//]: # (-.- .unlisted .unnumbered)

Now, we are ready to estimate the OLS regression model and at the same time we will introduce inference in this model.

## Single coefficient hypotheses, t-tests

We use t-tests to test simple coefficient restrictions on regression coefficients. Let's model `lwage` as a function of experience and the log of the husband's wage (`log_huswage'`).

In [33]:
# Defining the model
model = smf.ols('lwage ~ exper + log_huswage', data=mydata)
reg_ex1 = model.fit()

# Displaying the regression results
print(reg_ex1.summary())

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.062
Model:                            OLS   Adj. R-squared:                  0.058
Method:                 Least Squares   F-statistic:                     14.04
Date:                Sat, 25 Jan 2025   Prob (F-statistic):           1.25e-06
Time:                        21:50:47   Log-Likelihood:                -454.41
No. Observations:                 428   AIC:                             914.8
Df Residuals:                     425   BIC:                             927.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept       0.3590      0.175      2.047      

As you can see, this output contains $t$-statistics and their associated p-values. These test statistics and their p-values are associated with the following hypothesis test:

$$
H_0: \beta_i = 0; \quad H_A: \beta_i \neq 0
$$

Here, $β_i$ represents the $i$-th unknown population parameter. If you want to test any other hypothesis (rather than the two-sided test for equality to 0), you will need to access the regression output to calculate the relevant test statistic.

For now, let us consider the following hypothesis:

$$
H_0: \beta_{\text{exper}} = 0.01; \quad H_A: \beta_{\text{exper}} > 0.01
$$

To perform this test, we need to calculate the test statistic:

$$
t\text{-stat} = \frac{\hat{\beta}_i - 0.01}{\text{se}(\hat{\beta}_i)}
$$


Let's now investigate the significance of individual coefficients using t-tests.

In [35]:
print(reg_ex1.summary().tables[1])

                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept       0.3590      0.175      2.047      0.041       0.014       0.704
exper           0.0168      0.004      3.973      0.000       0.009       0.025
log_huswage     0.3033      0.078      3.884      0.000       0.150       0.457


Noting that the results for the `exper` variables are in the 2nd row. We can perform a one-sided t-test where the null hypothesis is that the coefficient of `exper` is equal to 0.01, and the alternative hypothesis is that it is greater than 0.01. 

In [37]:
# Hypothesis: H0: beta_exper = 0.01; HA: beta_exper > 0.01
beta_hat = reg_ex1.params['exper']
se_beta_hat = reg_ex1.bse['exper']
t_test = (beta_hat - 0.01) / se_beta_hat

# Display the t-test result
print(f'T-test statistic: {t_test}')

T-test statistic: 1.613707572362997


As it turns out the value of this t-test is **1.6137**. We do not know the p-value for this particular test yet. We will turn to this shortly.

## Multiple coefficient hypotheses, F-tests

Suppose we're interested in determining the impact of adding two new variables—`age` and `educ`—to our regression model that already includes `exper` and `log_huswage`. Conducting an F-test allows us to evaluate whether these additions significantly improve the model statistically.

[//]: # (-.- .tabset)

#### Step 1: Specify Both Models
First, we need to define both the restricted model (without the additional variables) and the unrestricted model (with the additional variables).

In [39]:
import statsmodels.formula.api as smf

# Restricted model (without 'age' and 'educ')
restricted_model = smf.ols('lwage ~ exper + log_huswage', data=mydata).fit()

# Unrestricted model (with 'age' and 'educ')
unrestricted_model = smf.ols('lwage ~ exper + log_huswage + age + educ', data=mydata).fit()

#### Step 2: Conduct the F-test

We then use the F-test to compare these models. The F-test examines if the explained variance in the unrestricted model is significantly greater, in proportion to the added variables, than in the restricted model.

In [41]:
from statsmodels.stats.anova import anova_lm

# Conducting the F-test
f_test_result = anova_lm(restricted_model, unrestricted_model)
print(f_test_result)

   df_resid         ssr  df_diff    ss_diff         F        Pr(>F)
0     425.0  209.488800      0.0        NaN       NaN           NaN
1     423.0  188.023144      2.0  21.465656  24.14589  1.175388e-10


As you can indicate, the output from `anova_lm` includes an F-statistic and an associated p-value. A significant p-value (typically <0.05) suggests that the additional variables (`age` and `educ`) contribute meaningfully to the model, thus justifying their inclusion. More specifically from the output of the F-test comparing the restricted and unrestricted models, you have the following key statistics:

- `df_resid`: This shows the degrees of freedom associated with the residuals from each model. For the unrestricted model (1), it is 423, indicating it has more predictors than the restricted model (0), which has 425.
  
- `ssr (sum of squared residuals)`: This indicates the total deviation of the response values from the fitted values. For the unrestricted model, it is lower (188.023144) compared to the restricted model (209.488800), suggesting a better fit as it explains more variability.

  
- `df_diff`: The difference in degrees of freedom between the models, which is 2. This corresponds to the two additional variables (age and educ) included in the unrestricted model.

  
- `ss_diff`: The difference in the sum of squared residuals between the restricted and unrestricted models, which is 21.465656. This quantifies how much more of the variance in the dependent variable is explained by including age and educ.

  
- `F`: The computed value of the F-statistic, 24.14589, measures the joint effect of age and educ on the model.

  
- `Pr(>F)`: The p-value associated with the F-test, 1.175838e-10, is extremely small, far less than any conventional significance level (e.g., 0.05, 0.01).


The very small p-value indicates that you can reject the null hypothesis that `age` and `educ` have no effect on `lwage`. In other words, adding these variables to the model provides a statistically significant improvement in model fit. Moreover, the increase in the F-statistic and the corresponding low p-value strongly suggest that including age and educ improves the model's ability to explain the variability in `lwage`.


[//]: # (-.- .unlisted .unnumbered)

### Calculating p-values

When calculating p-values for test statistics, several components are crucial:

- **Test Statistic Value**: You need the test statistic value you've calculated based on your sample data.
- **Distribution of the Test Statistic**: Understand the theoretical distribution of the test statistic under the null hypothesis. This will often depend on the test's assumptions about the data (e.g., normality).
- **Type of Test**: Determine whether you're conducting a two-tailed, left-tailed, or right-tailed test based on your hypothesis.

For a t-test example where the hypothesis is 
$$ H_0: \beta_{\text{exper}} = 0.01$$ 
versus 
$$ H_A: \beta_{\text{exper}} > 0.01 $$


you would perform a right-tailed test.

To calculate the p-value for a right-tailed test using a t-distribution, in Python, you can employ the `scipy.stats` library, which provides a comprehensive toolkit for statistical testing and distribution analysis. This library calculates probabilities from a t-distribution (probability under the t-distribution with 425 degrees of freedom to the left of the test-statistic t_test). The degrees of freedom is the number of observations minus the number of estimated regression coefficients (here 428-3 = 425). In our scenario, we are testing whether the coefficient of `exper` is significantly greater than 0.01. Let's calculate the p-value using Python for out t-statistic, where we are interested in the probability that the true value is greater than the calculated t-statistic under the null hypothesis.

In [43]:
import scipy.stats as stats

# Assuming t_test is your calculated t-statistic, and df_resid is your degrees of freedom
t_test = 1.613  
df_resid = 425  

# Calculating the p-value for a right-tailed test using the survival function, which is equivalent to 1 - CDF
p_value = stats.t.sf(t_test, df=df_resid)

# Print the p-value
print(f"P-value for right-tailed test: {p_value:.3f}")

P-value for right-tailed test: 0.054



In Python, the approach to obtaining p-values for various statistical tests, including F-tests, similarly involves using appropriate distribution-specific functions provided by libraries such as `scipy.stats`. This suite of functions allows you to work with a broad spectrum of distributions, offering methods for both calculating critical values and determining p-values. For a list see [here](https://docs.scipy.org/doc/scipy/reference/stats.html) 

## Robust Inference in Python

In econometric analyses, tests like t-tests and F-tests typically rest on certain assumptions about the data, such as homoskedasticity, which assumes that the error terms ($u_i$) have constant variance across all observations. When these assumptions are not met—particularly with heteroskedastic data—the standard error calculations can be biased, leading to incorrect inferences.


### Heteroskedasticity Robust t-tests in Python

In econometric analysis, particularly when addressing heteroskedasticity in regression models, it's crucial to use robust standard errors to ensure accurate inference. Python provides tools for this, specifically through the `statsmodels` library. By utilisizing the `statsmodels` library, we specify a covariance type when you retrieve the model results. Let's explore how this is implemented:

[//]: # (-.- .tabset)

##### Step 1: Fit Your Regression Model
Fit your model using ordinary least squares (OLS) or another appropriate method provided by `statsmodels`.


In [108]:
# Fit an OLS model
model = smf.ols('lwage ~ exper + log_huswage', data=mydata)
results = model.fit()

##### Step 2: Apply Robust Standard Errors

After fitting the model, apply robust standard errors using the `get_robustcov_results` method, specifying the type of heteroskedasticity-consistent (HC) standard errors.

In [110]:
# Use HC1 robust standard errors
robust_results = results.get_robustcov_results(cov_type='HC1')
print(robust_results.summary())

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.062
Model:                            OLS   Adj. R-squared:                  0.058
Method:                 Least Squares   F-statistic:                     13.62
Date:                Sat, 25 Jan 2025   Prob (F-statistic):           1.84e-06
Time:                        22:12:56   Log-Likelihood:                -454.41
No. Observations:                 428   AIC:                             914.8
Df Residuals:                     425   BIC:                             927.0
Df Model:                           2                                         
Covariance Type:                  HC1                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept       0.3590      0.180      1.998      

Here, the `cov_type='HC1'` option is one among several available choices (e.g., `HC0`, `HC1`, `HC2`, `HC3`, `HC4`). Each type adjusts for heteroskedasticity in slightly different ways:
- `HC0`: The original White’s estimator, providing a simple correction.
- `HC1`: Adjusts the HC0 standard errors for degrees of freedom, similar to what is done in STATA and often considered a more reliable estimator under small samples.
- `HC2`: Adds further bias adjustment to the standard errors, useful in the presence of highly influential observations.
- `HC3`: More aggressively adjusts for leverage points, and is considered one of the most conservative.

As you can see, the output from `robust_results.summary()` includes the (unchanged) regression coefficients along with their new robust standard errors, t-values, and p-values.

[//]: # (-.- .unlisted .unnumbered)

In Python, we can create a function to display regression tables incorporating robust standard errors using the `stargazer_hc` function, which you can import into your working directory. This file is available for download from here: [Stargazer_hc.py](https://github.com/SimoKF/Drafting-Python-resources-for-ECLR/blob/main/Stargazer_hc.py). You should save this file into your working directory. You can then make this function available to your code using this line. This function, `stargazer_HC`, will adjust the standard errors of given models to be robust against heteroskedasticity and then generate regression tables in the specified format.

Since you've set your working directory to where the `Stargazer_hc.py` file is located, Python should be able to import it directly. However, to ensure Python always finds your module regardless of the current working directory, you can append the module’s directory to Python’s path list using the `sys` module:

Once the file is in the correct location, you can import the function defined in this file. The file is defined as Stargazer_hc.py and it contains functions named `display_model_results`. 

In [97]:
import os
print("Current Working Directory:", os.getcwd())

Current Working Directory: C:\Users\Admin\Desktop\Pycode\ECLR\data


In [99]:
import Stargazer_hc
print(Stargazer_hc.__file__)

C:\Users\Admin\Stargazer_hc.py


In [103]:
with open('C:/Users/Admin/Desktop/Pycode/ECLR/data/Stargazer_hc.py', 'r') as file:
    print(file.read())

from stargazer.stargazer import Stargazer
from IPython.display import HTML
import statsmodels.formula.api as smf
import pandas as pd

def display_model_results(data, formula, cov_type='HC1'):
    """
    Fit an OLS model and display results using Stargazer and print robust standard errors.
    
    Args:
        data (pd.DataFrame): DataFrame containing the data.
        formula (str): A formula string for the regression model.
        cov_type (str): Type of covariance matrix estimator for robust SEs.
    """
    # Fit the OLS model
    model = smf.ols(formula, data=data).fit()

    # Create a Stargazer object from the fitted model
    stargazer = Stargazer([model])

    # Display the formatted table in Jupyter Notebook
    display(HTML(stargazer.render_html()))

    # Calculate and display robust standard errors separately
    robust_results = model.get_robustcov_results(cov_type=cov_type)
    robust_se = robust_results.bse

    print("Robust Standard Errors:")
    print(robust_se)



Since you have already computed the robust results, you can directly pass these results to the `stargazer_HC` function to format and display them. Here is how you can do it:

In [105]:
import importlib
importlib.reload(Stargazer_hc)
from Stargazer_hc import display_model_results

In [112]:
# Display the model results using the custom function
display_model_results(mydata, 'lwage ~ exper + log_huswage')

Robust Standard Errors:
[0.17970218 0.00427371 0.08012121]


As you can indicate from the `Stargazer` output provided, he standard error in parenthesis report the same standard errors as calculated  from `statsmodels` using `get_robustcov_results(cov_type='HC1')`.

So, if calculating heteroskedasticity robust standard errors is something you do frequently it is best to use the `Stargazer_H`C function.

## Heteroskedasticity and autocorrelation Robust t-tests

When error terms exhibit autocorrelation and potentially heteroskedasticity, similar concerns apply as in R, and it becomes necessary to utilize robust standard error estimators, such as the **Newey-West estimator**. In Python, using the `statsmodels` package, this can be accomplished with functions from the `statsmodels.stats.sandwich_covariance module`. 

> The equivalent function to R's `vcovHAC()` is `statsmodels.stats.sandwich_covariance.cov_hac`.

[//]: # (-.- .tabset)


#### Step 1: Model Fitting

We have already fitted a model using OLS. Here's the basic setup, for context:

In [134]:
from statsmodels.regression.linear_model import OLS
from statsmodels.stats.sandwich_covariance import cov_hac
import statsmodels.api as sm

# Assuming 'mydata' DataFrame is already loaded and ready
X = sm.add_constant(mydata[['exper', 'log_huswage']])  # add a constant to the model
y = mydata['lwage']

model = OLS(y, X)
results = model.fit()

#### Step 2: Apply Newey-West Standard Errors

We will now apply Newey-West standard errors to the model you've fitted. This involves calculating the robust covariance matrix using `cov_ha`c from `statsmodels.stats.sandwich_covariance`.

In [150]:
# Calculate the Newey-West standard errors using cov_hac
# Note: `maxlags` should be directly supplied, not in a dictionary if causing an issue
nw_cov = cov_hac(results, maxlags=1)

TypeError: cov_hac_simple() got an unexpected keyword argument 'maxlags'

When executing this code, you might encounter the error: "TypeError: cov_hac_simple() got an unexpected keyword argument 'maxlags'". Don't let this setback deter you! This is a perfect opportunity to deepen your understanding and sharpen your problem-solving skills.Start by exploring the [statsmodels documentation](https://www.statsmodels.org/dev/generated/statsmodels.stats.sandwich_covariance.cov_hac.html), which offers detailed insights into the Newey-West estimator, known for its robustness against heteroscedasticity and autocorrelation. Additionally, engaging with the community on platforms like [stack overflow](https://stackoverflow.com/questions/23420454/newey-west-standard-errors-for-ols-in-python) can provide practical advice and innovative solutions. Dive into these resources to navigate through the complexities and enhance your analytical prowess!

Therefore, after fitting the model, use `get_robustcov_results` to adjust for autocorrelation and potential heteroskedasticity using Newey-West estimators.

In [182]:
hac_model = model.get_robustcov_results(cov_type='HAC', maxlags=1)

#### Step 3: Display Results with HAC Standard Errors
Use the robust covariance matrix to display the results, including coefficients, Newey-West standard errors, t-statistics, and p-values.

In [184]:
print(hac_model.summary())

                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.062
Model:                            OLS   Adj. R-squared:                  0.058
Method:                 Least Squares   F-statistic:                     13.86
Date:                Sat, 25 Jan 2025   Prob (F-statistic):           1.47e-06
Time:                        23:07:54   Log-Likelihood:                -454.41
No. Observations:                 428   AIC:                             914.8
Df Residuals:                     425   BIC:                             927.0
Df Model:                           2                                         
Covariance Type:                  HAC                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           0.3590      0.181      1.983      

[//]: # (-.- .unlisted .unnumbered)

Again, If you need to present regression analysis with autocorrelation and heteroskedasticity-robust standard errors (also known as Newey-West standard errors) in Python, you can use a specifically developed script called `Stargazer_hac.py`,  which you can download from here: [Stargazer_HAC.R](https://github.com/SimoKF/Drafting-Python-resources-for-ECLR/blob/main/Stargazer_hac.py). What you need to do is:

1. **Download and Save the Function**:

- First, obtain the `Stargazer_hac.py` script. You should download this file and save it into your working directory. If the file is available online, you might use a direct link to download it or manually place it in the appropriate folder.
2. Make the Function Available:

- In Python, unlike in R, you don't need to use a `source` function. Instead, ensure the script is in your current working directory (or a directory recognized by your Python environment). You can also explicitly append the file's directory to your Python path if it's not in the current directory:

In [220]:
import os
print("Current Working Directory:", os.getcwd())

Current Working Directory: C:\Users\Admin\Desktop\Pycode\ECLR\data


In [222]:
import Stargazer_hac
print(Stargazer_hac.__file__)

C:\Users\Admin\Desktop\Pycode\ECLR\data\Stargazer_hac.py


In [224]:
with open('C:/Users/Admin/Desktop/Pycode/ECLR/data/Stargazer_hac.py', 'r') as file:
    print(file.read())

from stargazer.stargazer import Stargazer
from IPython.display import HTML
import statsmodels.formula.api as smf
import pandas as pd

def display_model_results_hac(data, formula, maxlags=1):
    """
    Fit an OLS model and display results using Stargazer and print Newey-West HAC standard errors.
    
    Args:
        data (pd.DataFrame): DataFrame containing the data.
        formula (str): A formula string for the regression model.
        maxlags (int): Maximum number of lags to use for the HAC estimator.
    """
    # Fit the OLS model
    model = smf.ols(formula, data=data).fit()

    # Apply Newey-West HAC standard errors
    hac_model = model.get_robustcov_results(cov_type='HAC', maxlags=maxlags)

    # Create a Stargazer object from the fitted model with HAC standard errors
    stargazer = Stargazer([model])

    # Display the formatted table in Jupyter Notebook
    display(HTML(stargazer.render_html()))

    # Print Newey-West standard errors separately
    print("Newey-West 

And once you have done that you can call `Stargazer_hac` to generate the usual regression output but with autocorrelation- and heteroskedasticity-robust standard errors (aka newey-West standard errors).

In [226]:
import importlib
importlib.reload(Stargazer_hac)
from Stargazer_hac import display_model_results_hac

In [228]:
# Display the model results using the custom function
display_model_results_hac(mydata, 'lwage ~ exper + log_huswage')

Newey-West HAC Standard Errors:
[0.18104644 0.00420329 0.08258127]


Of course, in this example, using cross-sectional data, it is impossible to have autocorrelated standard errors. However, Python's `statsmodels` simply executes the instructions as provided without checking whether the data context makes autocorrelation plausible. It applies the Newey-West HAC standard error adjustments regardless!

## Robust F-tests


F-tests are also based on the assumption of homoskedasticity. Fortunately, addressing potential heteroskedasticity works similarly to how it is handled for robust standard errors. In Python, we can use the `statsmodels` library to specify and compute robust variance-covariance matrices and feed them into a hypothesis testing function.

We begin by specifying the same larger model as used in the standard inference section.

In [271]:
# Fit the larger model
model = smf.ols('lwage ~ exper + np.log(huswage) + age + educ', data=mydata).fit()

At the core of tackling the heteroskedasticity issue is again the calculation of the correct variance-covariance matrix, which is then fed into the lht function via the `get_robustcov_results` option.

In [273]:
# Use get_robustcov_results to calculate robust variance-covariance matrix (HC1)
robust_model = model.get_robustcov_results(cov_type='HC1')

The, you can perfrom the **robust F-test** for the specified null hypothesis

In [281]:
# Check model parameters
print("Model parameters:", model.params.index)

# Define the restriction matrix
restriction_matrix = np.array([
    [0, 0, 0, 1, 0],  # Corresponds to "age = 0"
    [0, 0, 0, 0, 1]   # Corresponds to "educ = 0"
])

# Perform Wald test (equivalent to lht in R)
wald_test = robust_model.wald_test(restriction_matrix, scalar=False)

# Display the results
print(wald_test)


Model parameters: Index(['Intercept', 'exper', 'np.log(huswage)', 'age', 'educ'], dtype='object')
<F test: F=array([[29.5604912]]), p=9.626906508172405e-13, df_denom=423, df_num=2>



We can see that the **F-statistic (29.6)** is very large, and the **p-value (9.6e-13)** is extremely small, indicating strong evidence against the null hypothesis. The result remains unchanged, an overwhelming rejection of the null hypothesis that both age and educ are irrelevant. This means:

- Both variables (`age` and `educ`) significantly contribute to explaining the variation in `lwage` (log wage).
- The rejection of the null hypothesis suggests that omitting these variables from the model would lead to a significant loss of explanatory power.

This exercise is part of the [ECLR](https://datasquad.github.io/ECLR/) page.